src/main/thrift/parquet.thrift

```diff
@@ -977,6 +1038,25 @@ struct ColumnIndex {
 
   /** A list containing the number of null values for each page **/
   5: optional list<i64> null_counts
+
+  /**
+   * Contains repetition level histograms for more details) for each page
+   * concatenated together.  The repetition_level_histogram field on
+   * SizeStatistics contains more details.
+   *
+   * When present the length should always be (number of pages *
+   * (max_repetition_level + 1)) elements in size.
+   *
+   * Element 0 is the first element of the histogram for the first page.
+   * Element (max_repetition_level + 1) is the first element of the histogram
+   * for the second page.
```

Comment:
> Column-major vs row-major terminology confused me, so I documented one approach. If there isn't consensus on the ordering here, lets please create a new thread.

I'll take a stab (and likely be wrong :wink:). I believe the discussion has assumed the histograms form a matrix with the row index being page number, the column index being the level. Assuming that, what you have defined would be the row major ordering, where elements of the same row are contiguous in memory, as in a C matrix. What @JFinis and @pitrou seem to prefer is the opposite, where elements of the same column are contiguous in memory, as in a Fortran matrix.

Reply:
Yeah, I think i got lost in whether column-major meant "page-major" or "histogram-major".  I'll let @JFinis and @pitrou chime in if they want this reversed.

Reply:
Well, I guess I'm confused by the terminology as well. Perhaps we can drop "row-major" and "column-major" entirely when discussing this? :-)

Personally, what I mean is that the levels of a given page should be contiguous in memory. I suppose that can be called "page-major" but I might be mistaken.

(if it was a 2D array or list, you would index first by page and then by level)

Reply:
I think maybe we should move away from the terminology and just assess if the comment matches what you want (which I think it does).

The other "major" option would be something like `Element 0 is the first element of the histogram for the first page.  Element 1 is the first element of the histogram from the second page.  The element at index `num_pages` is the second element of the histogram for the first page`

Reply:
> Agreed on moving away from the terminology "major" terminology and just assess if the comment matches what you want (which I think it does).

Yes, it does.

That said, I don't have a strong opinion on this anymore. I got confused by the "column-major" terminology and mistakenly assumed that Parquet columns were involved. Actually they're not :-)

Reply:
+1 for the current wording.

Reply:
I'll also point out that if we are very concerned about eacking out the most efficiency for reads then probably the best organization here is having levels grouped by pages, and having each entry be the  cumulative sum:

`[<page 0, histogram[0]>, <page 1, histogram[0]>, <page 0, histogram[0] + histogram[1]>, <page 1, histogram[0] + histogram[1]> ]`. 

this allows comparing, all not null doing a vectorized subtraction and checking for positive elements of the results.  I'm happy to change this if we don't think this is too much complexity.

Reply:
So you are proposing transposing the 2D interpretation of the array as currently defined, so that the histogram data for a given level is contiguous in memory, and further proposing that each level is the inclusive sum of the preceding levels.  So the last "column" would be the same histogram as in the column chunk's `SizeStatistics`. Is this correct?

I'm fine with the transposition if that makes the filtering use case easier. But since my primary concern is the size estimation, I'd prefer not having to do the extra work of computing deltas to figure out value counts. But at least with your proposal the counts are inclusive, so I wouldn't have to look outside the column index to get the information I want (unlike using the offset index to get page row counts, where you have to look at the row group's `num_rows` to get the row count for the last page).

Reply:
I'm listing an option, I agree there is tension between size estimation use-case and filtering use-case, it really depends what we want to favor (the plus side of favoring size estimation is it a more obvious representation).

Reply:
> I'm listing an option,

Understood. And I'm just voicing a preference :wink: I don't need to squeeze _every_ last microsecond out of size estimation, so am happy to yield to those with more pressing performance constraints.

Reply:
I would expect the per level representation to be slightly superior, as it is more useful for filtering. Filtering is a process that might lead to most pages being skipped, so the overall query time might be super short in this case. The most extreme case would be a point look-up where only a single row in a single page survives the filters. In this case, the performance of actually performing the filtering on the page index might have a measurable impact.

In contrast, for the size-estimation case, we're estimating the size because we're planning to read the page. This reading will take orders of magnitude longer, so it is not too important to avoid every possible cache-miss in this case.

That being said, we're talking about micro optimizations here. Even though my gut feeling is that the other ordering would be superior, I don't mind this order. We're not creating hundreds of lists anymore, that's the most important point for performance.

Reply:
Speaking of data, as will surprise no one, the current set of changes has a huge impact on the column index size (at least for the cases with thousands of pages with short histograms). I can post an update to the index size tables if there's interest.